File: Нахождение цикла отрицательного веса за O(n·m).cpp
#include <bits/stdc++.h>

using namespace std;

const int INF = 1e9;

vector<int> solve(const vector< tuple<int,int,int> > &e, int n, int m)
{
    vector<int> d (n,INF);
    vector<int> p (n,-1);
    int x;
    for (int i = 0;i < n;++i)
    {
        x = -1;
        for (int j = 0;j < m;++j)
            if (d[get<1>(e[j])] > d[get<0>(e[j])] + get<2>(e[j]))
            {
                d[get<1>(e[j])] = (-INF > d[get<0>(e[j])] + get<2>(e[j])) ? -INF : d[get<0>(e[j])] + get<2>(e[j]);
                p[get<1>(e[j])] = get<0>(e[j]);
                x = get<1>(e[j]);
            }
    }
    if (x == -1)
        return {};
    else
    {
        int y = x;
        for (int i = 0;i < n;++i)
            y = p[y];
        vector<int> path;
        int cur = y;
        while (true)
        {
            path.push_back(cur);
            if (cur == y && path.size() > 1)
                break;
            cur = p[cur];
        }
        reverse(path.begin(), path.end());
        return path;
    }
}

/*
-e - массив ребер.
 n - количество вершин.
 m - количество ребер.
-Граф хранить в виде списка ребер.
 Элемент в массиве edges вида (i,j,w) означает, что можно попасть из i в j за цену w.
-Функция solve возвращает массив, который является циклом отрицательного веса если он есть, 
 иначе пустой массив.
-Нумерация с 0.
(Алгоритм Форда - Беллмана)
*/

File: Нахождение цикла отрицательного веса за O(n·m).py
INF = 10**9
def solve(e:list,n:int,m:int) -> list:
    d = [0]*n
    p = [-1]*n
    for i in range(n):
        x = -1
        for j in range(m):
            if (d[e[j][1]] > d[e[j][0]] + e[j][2]):
                d[e[j][1]] = -INF if (-INF > d[e[j][0]] + e[j][2]) else d[e[j][0]] + e[j][2]
                p[e[j][1]] = e[j][0]
                x = e[j][1]
    if (x == -1):
        return [];
    else:
        y = x
        for i in range(n):
            y = p[y]
        path = []
        cur = y
        while (True):
            path.append(cur)
            if (cur == y and len(path) > 1):
                break;
            cur = p[cur]
        path.reverse()
        return path;

'''
-e - массив ребер.
 n - количество вершин.
 m - количество ребер.
-Граф хранить в виде списка ребер.
 Элемент в массиве edges вида (i,j,w) означает, что можно попасть из i в j за цену w.
-Функция solve возвращает массив, который является циклом отрицательного веса если он есть, 
 иначе пустой массив.
-Нумерация с 0.
(Алгоритм Форда - Беллмана)
'''
